app/services/slot_builder.rb

```diff
@@ -149,10 +149,7 @@ def busy_times_for(range, plage_ouverture, off_days = [])
       end
 
       def busy_times_from_rdvs(range, plage_ouverture)
-        rdv_starts_in_range = plage_ouverture.agent.rdvs.not_cancelled.where(starts_at: range)
-        rdv_ends_in_range = plage_ouverture.agent.rdvs.not_cancelled.where(ends_at: range)
-        rdv_over_range = plage_ouverture.agent.rdvs.not_cancelled.where("starts_at <= ?", range.begin).where("ends_at >= ?", range.end)
-        rdv_starts_in_range.or(rdv_ends_in_range).or(rdv_over_range).map do |rdv|
+        plage_ouverture.agent.rdvs.not_cancelled.where("starts_at < ?", range.end).where("ends_at > ?", range.begin).map do |rdv|
```

Comment:
Ça me parait raisonnable! Tant qu’on y est, autant utiliser [les tsrange et l’opérateur && de postgresql](https://www.postgresql.org/docs/current/functions-range.html), puisqu’on a indexé les rdv sur `tsrange(starts_at, ends_at, '[)'::text)`. 

Ça ferait 
```suggestion
        plage_ouverture.agent.rdvs.not_cancelled.where("tsrange(starts_at, ends_at, '[)') && tsrange(?, ?)", range.begin, range.end).map do |rdv|
```

Reply:
Dans un test rapide d’explain, on y gagne même un poil par rapport à la version avec les deux where distinct:

```
0> plage_ouverture.agent.rdvs.not_cancelled.where("starts_at < ?", range.end).where("ends_at > ?", range.begin).explain
=> EXPLAIN for: SELECT "rdvs".* FROM "rdvs" INNER JOIN "agents_rdvs" ON "rdvs"."id" = "agents_rdvs"."rdv_id" WHERE "agents_rdvs"."agent_id" = $1 AND "rdvs"."status" IN ($2, $3, $4) AND (starts_at < '2022-03-29 14:00:00') AND (ends_at > '2022-03-29 12:00:00') [["agent_id", 2], ["status", "unknown"], ["status", "waiting"], ["status", "seen"]]
                                                                                                    QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.56..12.62 rows=1 width=200)
   ->  Index Only Scan using index_agents_rdvs_on_agent_id_and_rdv_id on agents_rdvs  (cost=0.28..4.30 rows=1 width=8)
         Index Cond: (agent_id = '2'::bigint)
   ->  Index Scan using rdvs_pkey on rdvs  (cost=0.28..8.31 rows=1 width=200)
         Index Cond: (id = agents_rdvs.rdv_id)
         Filter: ((starts_at < '2022-03-29 14:00:00'::timestamp without time zone) AND (ends_at > '2022-03-29 12:00:00'::timestamp without time zone) AND (status = ANY ('{unknown,waiting,seen}'::rdv_status[])))
(6 rows)
```

```
0> plage_ouverture.agent.rdvs.not_cancelled.where("tsrange(starts_at, ends_at, '[)') && tsrange(?, ?)", range.begin, range.end).explain
=> EXPLAIN for: SELECT "rdvs".* FROM "rdvs" INNER JOIN "agents_rdvs" ON "rdvs"."id" = "agents_rdvs"."rdv_id" WHERE "agents_rdvs"."agent_id" = $1 AND "rdvs"."status" IN ($2, $3, $4) AND (tsrange(starts_at, ends_at, '[)') && tsrange('2022-03-29 12:00:00', '2022-03-29 14:00:00')) [["agent_id", 2], ["status", "unknown"], ["status", "waiting"], ["status", "seen"]]
                                                        QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.43..12.48 rows=1 width=200)
   Join Filter: (rdvs.id = agents_rdvs.rdv_id)
   ->  Index Scan using index_rdvs_on_tsrange_starts_at_ends_at on rdvs  (cost=0.15..8.17 rows=1 width=200)
         Index Cond: (tsrange(starts_at, ends_at, '[)'::text) && '["2022-03-29 12:00:00","2022-03-29 14:00:00")'::tsrange)
         Filter: (status = ANY ('{unknown,waiting,seen}'::rdv_status[]))
   ->  Index Only Scan using index_agents_rdvs_on_agent_id_and_rdv_id on agents_rdvs  (cost=0.28..4.30 rows=1 width=8)
         Index Cond: (agent_id = '2'::bigint)
(7 rows)
```

(Pour je ne sais quel raison, mes tests en local utilisent un index pour l’id de l’agent, et pas les tests que tu avais montré, mais à la limite on s’en fiche, PG peut très bien décider qu’un seq scan est plus pertinent dans certains conditions.)

Reply:
Merci ! C'est top.